File: robot_brain_system/decision.make/__ini__.cpp
import torch
import torch.nn as nn
import torch.optim as optim

# ----- Simple ANN -----
class NeurobotBrain(nn.Module):
    def __init__(self):
        super(NeurobotBrain, self).__init__()
        self.fc1 = nn.Linear(1, 10)  # Input: 1 sensor (distance)
        self.relu = nn.ReLU()
        self.fc2 = nn.Linear(10, 3)  # Output: 3 actions (FORWARD, LEFT, RIGHT)

    def forward(self, x):
        x = self.fc1(x)
        x = self.relu(x)
        x = self.fc2(x)
        return x

# ----- Example Training Step -----
brain = NeurobotBrain()
optimizer = optim.Adam(brain.parameters(), lr=0.01)
criterion = nn.CrossEntropyLoss()

# Example input: distance = 15cm
inputs = torch.tensor([[15.0]])  # shape: [batch, features]
target = torch.tensor([1])       # 0=FORWARD,1=LEFT,2=RIGHT

optimizer.zero_grad()
output = brain(inputs.float())
loss = criterion(output, target)
loss.backward()
optimizer.step()

print("ANN output:", output)

File: robot_brain_system/deep.rl/decision.loop/Ini.py
brain = StrategicANN()
reflex_snn = ReflexSNN()
optimizer = optim.Adam(brain.parameters(), lr=0.001)
criterion = nn.MSELoss()

actions = ["FORWARD", "LEFT", "RIGHT", "BACKWARD"]

try:
    cap = cv2.VideoCapture(0)
    while True:
        # ----- Read Arduino Sensors -----
        if ser.in_waiting:
            data = ser.readline().decode().strip()
            if data:
                # Example format: distance,imu_angle, lidar1,lidar2,...,lidar360
                parts = list(map(float, data.split(',')))
                distance = parts[0]
                imu = parts[1]
                lidar = parts[2:]
                sensor_input = process_lidar(lidar + [distance])  # 361 features
        
        # ----- Read Camera -----
        ret, frame = cap.read()
        if not ret:
            continue
        image_input = preprocess_camera(frame)
        
        # ----- ANN Strategic Decision -----
        ann_output = brain(image_input, sensor_input)
        action_idx = torch.argmax(ann_output).item()
        action = actions[action_idx]
        
        # ----- Reflexive SNN Control -----
        reflex_output = reflex_snn(sensor_input)
        # Convert to motor commands (0-180 degrees or speed)
        motor_command = reflex_output.detach().numpy()
        
        # ----- Send commands to Arduino -----
        ser.write((action + "\n").encode())
        
        # ----- Reward + Learning (placeholder for DQN/PPO) -----
        reward = 1 if distance > 15 else -1
        target = torch.zeros_like(ann_output)
        target[0, action_idx] = reward
        optimizer.zero_grad()
        loss = criterion(ann_output, target)
        loss.backward()
        optimizer.step()
        
        # ----- Swarm Communication -----
        position = [0,0,0]  # Example, replace with real odometry
        obstacles = lidar
        publish_state(position, obstacles)
        
        print(f"Action: {action} | Reward: {reward} | Motor spikes: {motor_command}")

except KeyboardInterrupt:
    cap.release()
    ser.close()
    print("Neurobot shut down")
